### ISSUE TITLE
Proposal: Add a #DESIGN_TIME preprocessor to help to write the designer specific codes

### ISSUE THREAD

**Issue by laolarou726** [AvaloniaUI/Avalonia#15313]
### Is your feature request related to a problem? Please describe.

Currently I/m using MSDI as my Dependency Injection container. However, because I injected all the required components in the VM's and UserControl's constructor. This will cause the AvaloniaUI's preview to stop working:

### Describe the solution you'd like

Introduce a preprocessor will easily solve this issue:

```c#

#if DESIGN_TIME
public RandomUserControl() // introduce an empty ctor for the previewer
{
    InitializeComponent();

    ...

    DataContext = new A();
}
#else
public RandomUserControl(A a, B b, C c)
{
    InitializeComponent();

    ...

    DataContext = a;
}
#endif


```

**Comment by maxkatz6** [comment]
Designer doesn't have a control over compilation process. It uses precompiled binaries. 
So, this feature is a bit problematic to implement.

**Comment by laolarou726** [comment]
> Designer doesn't have a control over compilation process. It uses precompiled binaries. 
> 
> So, this feature is a bit problematic to implement. 

Any possible ways to provide an empty ctor for the previewer in the design time? Currently all of my preview windows are dead because ctor with paras.

**Comment by maxkatz6** [comment]
@laolarou726 we technically can reuse Design.PreviewWith property.

Something like:
```csharp
public static class DesignData
{
    public static MyWindow Window => new MyWindow(DesignParam1, DesignParam2);
}
```

```xaml
<Window x:Class="MyWindow"
        Design.PreviewWith="{x:Static DesignData.MyWindow}">
    <Button />
</Window>
```

Where previewer will take Design.PreviewWith instance instead of creating object itself.
The problem is, though, with current designer/previewwith code  it will likely go into a recursion.

**Comment by timunie** [comment]
1. I think parameterless constructor is allowed to be marked as `internal`
2. We have `if (Design.IsDesingMode)` which can be used to have a code path to cover that
   a. Maybe above can be combined with `# IF DEBUG`, so the code is not in release mode

**Comment by maxkatz6** [comment]
Internal ctor should be supported even now.

**Comment by laolarou726** [comment]
Thanks! I'll try internal ctor with #if DEBUG check first

**Comment by laolarou726** [comment]
@maxkatz6 @timunie  Looks like the previewer still not working for the internal ctor:

```
System.Xaml.XamlException: “Unable to find public constructor for type XXXViewModel()
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/25716486/1f85f1bc-5c07-4bd5-a2dc-2654b5480cb0)
![image](https://github.com/AvaloniaUI/Avalonia/assets/25716486/d4dc00d8-0e40-4b5e-8ab4-a9707290b2e1)

**Comment by timunie** [comment]
yeah but for your ViewModel, check error message carefully!

**Comment by timunie** [comment]
@laolarou726 can you confirm the suggested solution works for you?

**Comment by laolarou726** [comment]
> @laolarou726 can you confirm the suggested solution works for you? 

Yes. It does work on some of my classes. But not all of them, I'm still trying to figure out why.
